### PR TITLE
Add Workspace Mappers

### PR DESCRIPTION
**Pull request by kwridan**
Part of #1765 

### Short description 📝

Recently as part of #1765 we identified a use case that required the introduction of a workspace wide mapper to apply modification to the top level workspace to enable the creation of new schemes.

### Solution 📦

Introduce `WorkspaceMapper`s that can be composed with the existing `ProjectMapper`s to enable the creation of a `AutomaticSchemesWorkspaceMapper` in #1765. 

### Implementation 👩‍💻👨‍💻

- [x] Introduce `WorkspaceMapper` + `ProjectWorkspaceMapper`
- [x] Integrate `WorkspaceMapper`s into the generation pipeline 

### Test Plan 🛠

Verify existing mappers continue to function _(most of this is already covered by acceptance tests)_, for example:

- Run `swift run tuist generate` within `fixtures/ios_app_with_custom_scheme`
- Verify workspace & projects are generated successfully
- Verify automaticly generates schemes (e.g. one per Target) still appear
- Run `swift run tuist generate` within `fixtures/ios_app_with_framework_and_resources`
- Verify workspace & projects are generates successfully
- Verify resouce accessors are still generated (`Assets+App.swift`)

### CODE REVIEW

File: Sources/TuistKit/GraphMappers/TuistWorkspaceIdentifierMapper.swift [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import Foundation
+import TuistCore
+import TuistSupport
+
+final class TuistWorkspaceIdentifierMapper: WorkspaceMapping {
```
Comment (kwridan) [inline Sources/TuistKit/GraphMappers/TuistWorkspaceIdentifierMapper.swift:5]: I took the opportunity to move this here as it nicely fits into this concept :)

File: Sources/TuistKit/GraphMappers/WorkspaceMapperProvider.swift [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    private let projectMapperProvider: ProjectMapperProviding
+    init(projectMapperProvider: ProjectMapperProviding = ProjectMapperProvider()) {
+        self.projectMapperProvider = projectMapperProvider
+    }
+
+    func mapper(config: Config) -> WorkspaceMapping {
+        var mappers: [WorkspaceMapping] = []
+
+        mappers.append(
+            ProjectWorkspaceMapper(mapper: projectMapperProvider.mapper(config: config))
```
Comment (kwridan) [inline Sources/TuistKit/GraphMappers/WorkspaceMapperProvider.swift:18]: We're able to integrate and compose any existing `Project` mappers via `ProjectWorkspaceMapper`

File: Sources/TuistKit/ProjectGenerator/ProjectGenerator.swift [hunk 3]
```diff
@@ -198,21 +197,42 @@ class ProjectGenerator: ProjectGenerating {
     }
 
     // swiftlint:disable:next large_tuple
-    private func loadProjectWorkspace(path: AbsolutePath) throws -> (Project, Graph, [SideEffectDescriptor]) {
-        // Load project
-        let (project, graph, sideEffects) = try loadProject(path: path)
+    private func loadProjectWorkspace(path: AbsolutePath) throws -> (Workspace, Project, Graph, [SideEffectDescriptor]) {
```
Comment (kwridan) [inline Sources/TuistKit/ProjectGenerator/ProjectGenerator.swift:200]: I couldn't find a neat way to remove the code de-dupe without impacting readability :/ 

The main differences between `loadProjectWorkspace` and `loadProject` now are:

- `Workspace` isn't created in `loadProject`
- Workspace mappers aren't applied in `loadProject`

File: Sources/TuistKit/GraphMappers/TuistWorkspaceIdentifierMapper.swift (hunk 1 above)
Comment (fortmarek) [reply Sources/TuistKit/GraphMappers/TuistWorkspaceIdentifierMapper.swift:5]: That's super nice! I'd just add a little bit of documentation as it's not really obvious from first looking at it what htis does 🙂

File: Sources/TuistKit/GraphMappers/WorkspaceMapperProvider.swift (hunk 2 above)
Comment (fortmarek) [reply Sources/TuistKit/GraphMappers/WorkspaceMapperProvider.swift:18]: also super nice 👏

**Review by fortmarek — APPROVED** [review]
This looks solid 👍  I think this does not have to be a draft anymore, so we can merge this soon-ish 🙂

File: Sources/TuistKit/GraphMappers/TuistWorkspaceIdentifierMapper.swift (hunk 1 above)
Comment (kwridan) [reply Sources/TuistKit/GraphMappers/TuistWorkspaceIdentifierMapper.swift:9]: Good idea, updated the doc comments for this mapper

**Review by pepibumur — APPROVED** [review]

**Comment by pepibumur** [comment]
I think we can merge after solving the git conflicts 😬

**Comment by pepibumur** [comment]
@all-contributors add @kwridan for code

**Pull request merged**